xds/internal/resolver/xds_resolver_test.go

```diff
@@ -547,21 +309,19 @@ func (s) TestResolverCloseClosesXDSClient(t *testing.T) {
 	// Override xDS client creation to use bootstrap configuration pointing to a
 	// dummy management server. Also close a channel when the returned xDS
 	// client is closed.
+	origNewClient := rinternal.NewXDSClient
 	closeCh := make(chan struct{})
-	origNewClient := newXDSClient
-	newXDSClient = func() (xdsclient.XDSClient, func(), error) {
+	rinternal.NewXDSClient = func() (xdsclient.XDSClient, func(), error) {
 		c, cancel, err := xdsclient.NewWithConfigForTesting(bootstrapCfg, defaultTestTimeout, defaultTestTimeout)
-		return c, func() {
+		return c, grpcsync.OnceFunc(func() {
```

Comment:
The `cancel` func returned by `xdsclient.NewWithConfigForTesting` is wrapped inside of a `grpcsync.OnceFunc`. So calling it any number of times is safe. But that is not the case with the cancel func that we are returning here.

And this becomes a real issue because the helper `buildResolverForTarget` calls `t.Cleanup(r.Close())` and this makes sense for most other tests which don't have to then manually call `defer r.Close()`. But here we have an explicit call to `r.Close()`. So, this means that we end up with two calls to `r.Close()`, which both end up calling `xdsClientClose` which is safe in the real code because of what I mentioned at the top of this comment.

Reply:
Alright sounds good.